core/unix/rseq_linux.c

```diff
@@ -0,0 +1,377 @@
+                if (ptrs < (byte **)ma->start || ptrs > (byte **)ma->end) {
+                    REPORT_FATAL_ERROR_AND_EXIT(
+                        RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                        get_application_pid(),
+                        RSEQ_PTR_ARRAY_SEC_NAME " is not in a loaded segment");
+                    ASSERT_NOT_REACHED();
+                }
+                /* We assume this is a full mapping and it's safe to read the data
+                 * (a partial map shouldn't make it to module list processing).
+                 * We do perform a sanity check to handle unusual non-relocated
+                 * cases (it's possible this array is not in a loaded segment?).
+                 */
+                byte *entry = *ptrs + entry_offs;
+                if (entry < ma->start || entry > ma->end) {
+                    REPORT_FATAL_ERROR_AND_EXIT(
+                        RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                        get_application_pid(),
+                        RSEQ_PTR_ARRAY_SEC_NAME "'s entries are not in a loaded segment");
+                    ASSERT_NOT_REACHED();
+                }
+                rseq_process_entry((struct rseq_cs *)entry, entry_offs);
+                ++ptrs;
+            }
+            break;
+        }
+        ++sec_hdr;
+    }
+    if (!found_array) {
+        sec_hdr = sec_hdr_start;
+        for (i = 0; i < elf_hdr->e_shnum; i++) {
+#define RSEQ_SEC_NAME "__rseq_cs"
+#define RSEQ_OLD_SEC_NAME "__rseq_table"
+            if (strcmp(strtab + sec_hdr->sh_name, RSEQ_SEC_NAME) == 0 ||
+                strcmp(strtab + sec_hdr->sh_name, RSEQ_OLD_SEC_NAME) == 0) {
+                /* There may be padding at the start of the section, so ensure we skip
+                 * over it.  We're reading the loaded data, not the file, so it will
+                 * always be aligned.
+                 */
+#define RSEQ_CS_ALIGNMENT (4 * sizeof(__u64))
+                struct rseq_cs *array = (struct rseq_cs *)ALIGN_FORWARD(
+                    sec_hdr->sh_addr + load_offs, RSEQ_CS_ALIGNMENT);
+                int j;
+                for (j = 0; j < sec_hdr->sh_size / sizeof(*array); ++j) {
+                    /* We require that the table is loaded.  If not, bail. */
+                    if (array < (struct rseq_cs *)ma->start ||
+                        array > (struct rseq_cs *)ma->end) {
+                        REPORT_FATAL_ERROR_AND_EXIT(
+                            RSEQ_BEHAVIOR_UNSUPPORTED, 3, get_application_name(),
+                            get_application_pid(),
+                            RSEQ_SEC_NAME " is not in a loaded segment");
+                        ASSERT_NOT_REACHED();
+                    }
+                    rseq_process_entry(array, entry_offs);
+                    ++array;
+                }
+                break;
+            }
+            ++sec_hdr;
+        }
+    }
+    res = true;
+rseq_process_module_cleanup:
```

Comment:
Do you think it's possible to re-factor this long function to shorter ones and no "gotos". This reminds me of old gcc code. Since we have the chance and writing this new and more readable? No hard opinion here but WDYT

Reply:
In C there are no destructors, so this pattern of a single goto target which contains cleanup code ends up being more readable and maintainable than embedding all the clean up code inside each if() above.  It is error-prone to have all these cleanup steps duplicated at each of 5 different places above: too easy to miss one, too easy to change in one place and forget to update the other 4, and duplicating makes the function much longer.  This is one pattern where goto improves maintainability in C.

For splitting the function: it is 150 lines, so medium-sized.  I think the only simple split would be moving the searches into a separate function.  I can do that.  Done.

Reply:
Ok.